File: bongus/core/binance_endpoints.py
"""Shared Binance endpoint and credential helpers.

Paper mode intentionally uses mainnet market data while suppressing real orders.
Testnet mode selects the official USD-M demo and Spot Testnet endpoints.
"""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any

ENDPOINT_MATRIX_PATH = (
    Path(__file__).resolve().parents[2] / "config" / "binance_endpoints_v1.json"
)


def _load_endpoint_matrix() -> dict[str, Any]:
    payload = json.loads(ENDPOINT_MATRIX_PATH.read_text(encoding="utf-8"))
    if payload.get("schema_version") != 1:
        raise RuntimeError("unsupported Binance endpoint-matrix schema")
    max_age = payload.get("planned_connection_max_age_seconds")
    if not isinstance(max_age, int) or not 0 < max_age < 24 * 60 * 60:
        raise RuntimeError("Binance connection renewal must be planned before 24 hours")
    environments = payload.get("environments")
    if not isinstance(environments, dict) or set(environments) != {"mainnet", "testnet"}:
        raise RuntimeError("Binance endpoint matrix must define mainnet and testnet")
    for environment in ("mainnet", "testnet"):
        venues = environments.get(environment)
        if not isinstance(venues, dict) or set(venues) != {"futures", "spot"}:
            raise RuntimeError(f"invalid Binance venue matrix for {environment}")
        for venue in ("futures", "spot"):
            endpoints = venues.get(venue)
            required = {
                "rest_base_url",
                "public_stream_ws_base_url",
                "market_stream_ws_base_url",
                "private_ws_base_url",
            }
            if not isinstance(endpoints, dict) or set(endpoints) != required:
                raise RuntimeError(f"invalid Binance endpoint set for {environment}/{venue}")
            if not str(endpoints["rest_base_url"]).startswith("https://"):
                raise RuntimeError("Binance REST endpoint must use HTTPS")
            for field in (
                "public_stream_ws_base_url",
                "market_stream_ws_base_url",
                "private_ws_base_url",
            ):
                if not str(endpoints[field]).startswith("wss://"):
                    raise RuntimeError("Binance stream endpoint must use WSS")
    return payload


_ENDPOINT_MATRIX = _load_endpoint_matrix()
PLANNED_CONNECTION_MAX_AGE_SECONDS = int(
    _ENDPOINT_MATRIX["planned_connection_max_age_seconds"]
)


def _endpoint(environment: str, venue: str, field: str) -> str:
    return str(_ENDPOINT_MATRIX["environments"][environment][venue][field])


MAINNET_FUTURES_REST_BASE_URL = _endpoint(
    "mainnet", "futures", "rest_base_url"
)
TESTNET_FUTURES_REST_BASE_URL = _endpoint(
    "testnet", "futures", "rest_base_url"
)
MAINNET_SPOT_REST_BASE_URL = _endpoint("mainnet", "spot", "rest_base_url")
TESTNET_SPOT_REST_BASE_URL = _endpoint("testnet", "spot", "rest_base_url")

MAINNET_FUTURES_STREAM_WS_BASE_URL = _endpoint(
    "mainnet", "futures", "market_stream_ws_base_url"
)
TESTNET_FUTURES_STREAM_WS_BASE_URL = _endpoint(
    "testnet", "futures", "market_stream_ws_base_url"
)
MAINNET_SPOT_STREAM_WS_BASE_URL = _endpoint(
    "mainnet", "spot", "market_stream_ws_base_url"
)
TESTNET_SPOT_STREAM_WS_BASE_URL = _endpoint(
    "testnet", "spot", "market_stream_ws_base_url"
)
MAINNET_FUTURES_PUBLIC_STREAM_WS_BASE_URL = _endpoint(
    "mainnet", "futures", "public_stream_ws_base_url"
)
TESTNET_FUTURES_PUBLIC_STREAM_WS_BASE_URL = _endpoint(
    "testnet", "futures", "public_stream_ws_base_url"
)
MAINNET_SPOT_PUBLIC_STREAM_WS_BASE_URL = _endpoint(
    "mainnet", "spot", "public_stream_ws_base_url"
)
TESTNET_SPOT_PUBLIC_STREAM_WS_BASE_URL = _endpoint(
    "testnet", "spot", "public_stream_ws_base_url"
)
MAINNET_FUTURES_PRIVATE_WS_BASE_URL = _endpoint(
    "mainnet", "futures", "private_ws_base_url"
)
TESTNET_FUTURES_PRIVATE_WS_BASE_URL = _endpoint(
    "testnet", "futures", "private_ws_base_url"
)
MAINNET_SPOT_PRIVATE_WS_BASE_URL = _endpoint(
    "mainnet", "spot", "private_ws_base_url"
)
TESTNET_SPOT_PRIVATE_WS_BASE_URL = _endpoint(
    "testnet", "spot", "private_ws_base_url"
)


def normalize_trading_mode(trading_mode: str | None = None) -> str:
    normalized = (trading_mode or os.getenv("TRADING_MODE", "paper")).strip().lower()
    return normalized if normalized in {"paper", "testnet", "live"} else "paper"


def uses_testnet(trading_mode: str | None = None) -> bool:
    return normalize_trading_mode(trading_mode) == "testnet"


def get_rest_base_urls(trading_mode: str | None = None) -> tuple[str, str]:
    if uses_testnet(trading_mode):
        return TESTNET_FUTURES_REST_BASE_URL, TESTNET_SPOT_REST_BASE_URL
    return MAINNET_FUTURES_REST_BASE_URL, MAINNET_SPOT_REST_BASE_URL


def get_stream_ws_base_urls(trading_mode: str | None = None) -> tuple[str, str]:
    if uses_testnet(trading_mode):
        return TESTNET_FUTURES_STREAM_WS_BASE_URL, TESTNET_SPOT_STREAM_WS_BASE_URL
    return MAINNET_FUTURES_STREAM_WS_BASE_URL, MAINNET_SPOT_STREAM_WS_BASE_URL


def get_public_ws_base_urls(trading_mode: str | None = None) -> tuple[str, str]:
    if uses_testnet(trading_mode):
        return (
            TESTNET_FUTURES_PUBLIC_STREAM_WS_BASE_URL,
            TESTNET_SPOT_PUBLIC_STREAM_WS_BASE_URL,
        )
    return (
        MAINNET_FUTURES_PUBLIC_STREAM_WS_BASE_URL,
        MAINNET_SPOT_PUBLIC_STREAM_WS_BASE_URL,
    )


def get_private_ws_base_urls(trading_mode: str | None = None) -> tuple[str, str]:
    if uses_testnet(trading_mode):
        return TESTNET_FUTURES_PRIVATE_WS_BASE_URL, TESTNET_SPOT_PRIVATE_WS_BASE_URL
    return MAINNET_FUTURES_PRIVATE_WS_BASE_URL, MAINNET_SPOT_PRIVATE_WS_BASE_URL


def resolve_binance_credentials() -> dict[str, str]:
    futures_api_key = os.getenv("BINANCE_API_KEY", "").strip()
    futures_api_secret = os.getenv("BINANCE_API_SECRET", "").strip()
    spot_api_key = os.getenv("BINANCE_SPOT_API_KEY", "").strip()
    spot_api_secret = os.getenv("BINANCE_SPOT_API_SECRET", "").strip()

    shared_api_key = futures_api_key or spot_api_key
    shared_api_secret = futures_api_secret or spot_api_secret

    return {
        "futures_api_key": shared_api_key,
        "futures_api_secret": shared_api_secret,
        "spot_api_key": spot_api_key or shared_api_key,
        "spot_api_secret": spot_api_secret or shared_api_secret,
    }

File: execution_engine/src/binance_endpoints.rs
use serde::Deserialize;
use std::collections::HashMap;

const ENDPOINT_MATRIX_JSON: &str = include_str!("../../config/binance_endpoints_v1.json");

#[derive(Debug, Clone, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct VenueEndpoints {
    pub rest_base_url: String,
    pub public_stream_ws_base_url: String,
    pub market_stream_ws_base_url: String,
    pub private_ws_base_url: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
struct EnvironmentEndpoints {
    futures: VenueEndpoints,
    spot: VenueEndpoints,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
struct EndpointMatrix {
    schema_version: u32,
    planned_connection_max_age_seconds: u64,
    environments: HashMap<String, EnvironmentEndpoints>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SelectedEndpoints {
    pub futures: VenueEndpoints,
    pub spot: VenueEndpoints,
    pub planned_connection_max_age_seconds: u64,
}

fn load_matrix() -> Result<EndpointMatrix, String> {
    let matrix: EndpointMatrix = serde_json::from_str(ENDPOINT_MATRIX_JSON)
        .map_err(|error| format!("parse shared Binance endpoint matrix: {error}"))?;
    if matrix.schema_version != 1 {
        return Err(format!(
            "unsupported Binance endpoint schema {}",
            matrix.schema_version
        ));
    }
    if matrix.planned_connection_max_age_seconds == 0
        || matrix.planned_connection_max_age_seconds >= 24 * 60 * 60
    {
        return Err("Binance connection renewal must be planned before 24 hours".to_string());
    }
    if matrix.environments.len() != 2
        || !matrix.environments.contains_key("mainnet")
        || !matrix.environments.contains_key("testnet")
    {
        return Err("Binance endpoint matrix must define mainnet and testnet".to_string());
    }
    for (environment, endpoints) in &matrix.environments {
        for (venue, values) in [("futures", &endpoints.futures), ("spot", &endpoints.spot)] {
            if !values.rest_base_url.starts_with("https://")
                || !values.public_stream_ws_base_url.starts_with("wss://")
                || !values.market_stream_ws_base_url.starts_with("wss://")
                || !values.private_ws_base_url.starts_with("wss://")
            {
                return Err(format!(
                    "invalid secure Binance endpoint for {environment}/{venue}"
                ));
            }
        }
    }
    Ok(matrix)
}

pub fn endpoints_for_mode(trading_mode: &str) -> Result<SelectedEndpoints, String> {
    let matrix = load_matrix()?;
    let environment = if trading_mode.eq_ignore_ascii_case("testnet") {
        "testnet"
    } else {
        "mainnet"
    };
    let endpoints = matrix
        .environments
        .get(environment)
        .ok_or_else(|| format!("missing Binance endpoint environment {environment}"))?;
    Ok(SelectedEndpoints {
        futures: endpoints.futures.clone(),
        spot: endpoints.spot.clone(),
        planned_connection_max_age_seconds: matrix.planned_connection_max_age_seconds,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shared_matrix_uses_official_testnet_endpoints_and_planned_renewal() {
        let endpoints = endpoints_for_mode("testnet").unwrap();
        assert_eq!(
            endpoints.futures.rest_base_url,
            "https://demo-fapi.binance.com"
        );
        assert_eq!(
            endpoints.futures.public_stream_ws_base_url,
            "wss://demo-fstream.binance.com/public"
        );
        assert_eq!(
            endpoints.futures.market_stream_ws_base_url,
            "wss://demo-fstream.binance.com/market"
        );
        assert_eq!(
            endpoints.spot.rest_base_url,
            "https://testnet.binance.vision"
        );
        assert_eq!(
            endpoints.spot.private_ws_base_url,
            "wss://ws-api.testnet.binance.vision/ws-api/v3"
        );
        assert_eq!(endpoints.planned_connection_max_age_seconds, 23 * 60 * 60);
    }

    #[test]
    fn paper_uses_mainnet_market_data_matrix() {
        assert_eq!(
            endpoints_for_mode("paper")
                .unwrap()
                .futures
                .market_stream_ws_base_url,
            "wss://fstream.binance.com/market"
        );
    }
}
